File: backbone-go/examples/clean-api-go/interfaces/http/middleware/logging_middleware.go
// Package middleware contains HTTP middleware
package middleware

import (
	"context"
	"net/http"
	"time"

	"github.com/freakjazz/backbone-go/infrastructure/logging"
	"github.com/google/uuid"
)

// ctxKey is an unexported type for context keys in this package.
// Using a private type prevents collisions with keys from other packages
// that happen to use the same string — a common source of subtle bugs.
type ctxKey uint8

const ctxRequestID ctxKey = 0

// RequestIDFromContext returns the request trace ID set by LoggingMiddleware.
// Returns an empty string if no ID is present (e.g. in unit tests).
func RequestIDFromContext(ctx context.Context) string {
	if v, ok := ctx.Value(ctxRequestID).(string); ok {
		return v
	}
	return ""
}

// LoggingMiddleware logs HTTP requests and responses.
// It also generates a UUID request trace ID and stores it in the context
// so all downstream handlers and use-cases can attach it to logs and error responses.
func LoggingMiddleware(logger *logging.EnhancedLogger) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			start := time.Now()

			requestID := uuid.New().String()
			ctx := context.WithValue(r.Context(), ctxRequestID, requestID)
			r = r.WithContext(ctx)

			reqLogger := logger.
				WithLayer("interfaces").
				WithComponent("HTTPMiddleware").
				WithMethod("LoggingMiddleware").
				WithContext(map[string]interface{}{
					"request_id": requestID,
				})

			reqLogger.Info("Incoming HTTP request", map[string]interface{}{
				"method":     r.Method,
				"path":       r.URL.Path,
				"query":      r.URL.RawQuery,
				"remote_ip":  r.RemoteAddr,
				"user_agent": r.UserAgent(),
			})

			rw := &responseWriter{ResponseWriter: w, statusCode: http.StatusOK}
			next.ServeHTTP(rw, r)

			reqLogger.Info("HTTP request completed", map[string]interface{}{
				"method":      r.Method,
				"path":        r.URL.Path,
				"status_code": rw.statusCode,
				"duration_ms": time.Since(start).Milliseconds(),
			})
		})
	}
}

type responseWriter struct {
	http.ResponseWriter
	statusCode int
}

func (rw *responseWriter) WriteHeader(code int) {
	rw.statusCode = code
	rw.ResponseWriter.WriteHeader(code)
}

File: examples/clean_api_python/interfaces/http/middleware/logging_middleware.py
"""Logging middleware for Flask"""
from flask import request, g
from functools import wraps
import time
import uuid
import sys
import os

# Add backbone to path
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../../../..')))

from backbone.infrastructure.logging import LoggerFactory, LogLevel


logger = LoggerFactory.create_layer_logger(
    "product-api",
    "interfaces",
    "HTTPMiddleware"
)


def logging_middleware(f):
    """Logs HTTP requests and responses"""
    @wraps(f)
    def decorated_function(*args, **kwargs):
        start = time.time()
        
        # Generate request ID
        request_id = str(uuid.uuid4())
        g.request_id = request_id
        
        # Log request
        logger.log(
            LogLevel.INFO,
            "Incoming HTTP request",
            extra_data={
                "method": request.method,
                "path": request.path,
                "query": request.query_string.decode(),
                "remote_ip": request.remote_addr,
                "user_agent": request.user_agent.string
            },
            context={"request_id": request_id}
        )
        
        # Call handler
        response = f(*args, **kwargs)
        
        # Calculate duration
        duration_ms = int((time.time() - start) * 1000)
        
        # Get status code
        if isinstance(response, tuple):
            status_code = response[1]
        else:
            status_code = 200
        
        # Log response
        logger.log(
            LogLevel.INFO,
            "HTTP request completed",
            extra_data={
                "method": request.method,
                "path": request.path,
                "status_code": status_code,
                "duration_ms": duration_ms
            },
            context={"request_id": request_id}
        )
        
        return response
    
    return decorated_function
